src/detect-dns-response.c

```diff
@@ -0,0 +1,307 @@
+                response_id++;
+                continue;
+            }
+
+            match = DetectEngineContentInspectionBuffer(de_ctx, det_ctx, s, engine->smd, NULL, f,
+                    buffer, DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE);
+            if (match) {
+                return DETECT_ENGINE_INSPECT_SIG_MATCH;
+            }
+            local_id++;
+            response_id++;
+        }
+    }
+
+    return DETECT_ENGINE_INSPECT_SIG_NO_MATCH;
+}
+
+static void DetectDnsResponsePrefilterTx(DetectEngineThreadCtx *det_ctx, const void *pectx,
+        Packet *p, Flow *f, void *txv, const uint64_t idx, const AppLayerTxData *_txd,
+        const uint8_t flags)
+{
+    SCEnter();
+
+    const PrefilterMpm *ctx = (const PrefilterMpm *)pectx;
+    const MpmCtx *mpm_ctx = ctx->mpm_ctx;
+    const int list_id = ctx->list_id;
+
+    uint32_t local_id = 0;
+    /* loop through each possible DNS response section */
+    for (uint8_t section = DNS_RESPONSE_QUERY; section < DNS_RESPONSE_MAX; section++) {
+        uint32_t response_id = 0;
+        /* loop through each record in section inspecting "name" and "rdata" */
+        while (1) {
+            struct DnsResponseGetDataArgs cbdata = { section, response_id, local_id };
+
+            /* extract resource record "name" */
+            InspectionBuffer *buffer =
+                    GetBuffer(det_ctx, flags, ctx->transforms, txv, &cbdata, list_id, false);
+            if (buffer == NULL) {
+                local_id++;
+                break;
+            }
+
+            if (buffer->inspect_len >= mpm_ctx->minlen) {
+                (void)mpm_table[mpm_ctx->mpm_type].Search(mpm_ctx, &det_ctx->mtc, &det_ctx->pmq,
+                        buffer->inspect, buffer->inspect_len);
+                PREFILTER_PROFILING_ADD_BYTES(det_ctx, buffer->inspect_len);
+            }
+
+            local_id++;
+            if (section == DNS_RESPONSE_QUERY) {
+                /* no rdata to inspect for query section, move on to next name entry */
+                response_id++;
+                continue;
+            }
+
+            /* extract resource record "rdata" */
+            cbdata.local_id = local_id;
+            buffer = GetBuffer(det_ctx, flags, ctx->transforms, txv, &cbdata, list_id, true);
+            if (buffer == NULL) {
+                local_id++;
+                response_id++;
+                continue;
+            }
+
+            if (buffer->inspect_len >= mpm_ctx->minlen) {
+                (void)mpm_table[mpm_ctx->mpm_type].Search(mpm_ctx, &det_ctx->mtc, &det_ctx->pmq,
+                        buffer->inspect, buffer->inspect_len);
+                PREFILTER_PROFILING_ADD_BYTES(det_ctx, buffer->inspect_len);
+            }
+            local_id++;
+            response_id++;
+        }
+    }
+}
+
+static void DetectDnsResponsePrefilterMpmFree(void *ptr)
+{
+    SCFree(ptr);
+}
+
+static int DetectDnsResponsePrefilterMpmRegister(DetectEngineCtx *de_ctx, SigGroupHead *sgh,
```

Comment:
Why do we need a specific stuff here ?


Do not we get prefilter for all sticky buffers from the generic engine ?

Reply:
> Do not we get prefilter for all sticky buffers from the generic engine ?

Are you asking? If so, I'm not sure.

Reply:
I think we do but you make me doubt...